### DISCUSSION TITLE
Add Background Image Resolution options for Dynamic Background Images

### DISCUSSION THREAD

**Discussion by ellengardt** [elementor/elementor#22750] [Feature Request]
### Prerequisites

- [ ] I have searched for similar features requests in both open and closed tickets and cannot find a duplicate.
- [X] The feature is still missing in the latest stable version of Elementor ( Elementor Pro )

### What problem is your feature request going to solve? Please describe.

Hi! In elementor 3.11, now possible to select the size of the background image. However, if the image is connected dynamically, its size cannot be set.

I create websites for people far from elementor and development, and use custom post types and settings to enter a large amount of data. Sometimes users upload 10 MB photos with a huge resolution for a block of 250*250 pixels

Therefore, the ability to specify the size that will be displayed in the background for dynamic bg-images is very necessary

### Describe the solution you'd like

The current solution for the static background is fine, I would like to see the same for dynamic images

### Additional context

![image](https://github.com/elementor/elementor/assets/35376009/d4900414-bf87-4e44-adc3-05c3415a5db7)![image](https://github.com/elementor/elementor/assets/35376009/a25d77c1-0a02-434f-8ff8-01ac30bc59f0)

**Comment by caspervoogt** [comment]
I just ran into the same thing. I have a post template that shows the featured image as a banner, but it is rendering as a full-size image instead of the image size I selected. I don't understand how this was not apparently tested for, when dynamic tags are part of Elementor core.

**Comment by contrastxxx** [comment]
I have the same problem, please fix, thank you

**Comment by steveindzine** [comment]
This is really important and must be fixed. I am using a column in a loop which sets the background image, it's now stuck at full size, I need to reduce this, but the only way is to use an image. I need this image to flexible in height and width which is why it's a background image.

**Comment by pumadog** [comment]
+1

**Comment by nicholaszein** [comment]
For future consultation.

This request was referenced in the following issues and discussions:

- https://github.com/elementor/elementor/issues/21359
- https://github.com/elementor/elementor/issues/22055
- https://github.com/orgs/elementor/discussions/22801
- https://github.com/elementor/elementor/issues/29714
- https://github.com/elementor/elementor/issues/30054

**Comment by stlhdcmptn** [comment]
Amazing to me that this hasnt been implemented yet.   It's really unfortunate that Elementor has shifted focus to profit maxing by way of hosting/ai.  I, and I'm sure most of the others would appreciate it if elementor focused on it's original product. Pushing buggy and incomplete updates to stable release environments with priority on ai/hosting features is NOT what we want.

**Comment by rtpHarry** [comment]
I dont know what they are referring to in the [comments they have posted when closing the other issues](https://github.com/elementor/elementor/issues/29714#issuecomment-2623102237). 

They say it has heavy impact on the site. Perhaps they were trying to do a "proper job" which lets you have multiple image size choices for responsive screen sizes? 

I couldn't just let the website go out with ~11mb of resources when I just needed a small image in a search results page, so I've written my own simple plugin.

I looked at the featured image tag and I looked at my code and I can't see what it could be that is so terrible.

Ultimately I have set it so that you can pick one resize, which is output as the plain url, nothing special supported for different responsive sizes, but that at least lets site builders stop it from serving the raw unoptimised full size image to the visitor.

The trick I found was to remove the ID from the returned object, which forced it to use the actual url - otherwise it would just load it from the ID and render the full size image still.

I've put my dev name in brackets so you can easily discern between the real version and mine - to phase it out in the future if they ever implement a better solution.

I'll do a proper release soon, but for now here is the code:

```php
<?php

namespace RunthingsElementorDynamicTags;

use Elementor\Controls_Manager;
use ElementorPro\Modules\DynamicTags\Tags\Base\Data_Tag;
use ElementorPro\Modules\DynamicTags\Module;

if ( ! defined( 'ABSPATH' ) ) {
    exit; // Exit if accessed directly.
}

/**
 * Featured Image with Size Selection Dynamic Tag
 */
class FeaturedImageSizeTag extends Data_Tag {
    /**
     * Get tag name
     */
    public function get_name() {
        return 'runthings-featured-image-size';
    }

    /**
     * Get tag title
     */
    public function get_title() {
        return __( 'Featured Image (runthings.dev)', 'runthings-dynamic-tag-image-size' );
    }

    /**
     * Get tag group
     */
	public function get_group() {
		return Module::POST_GROUP;
	}

    /**
     * Get tag categories
     */
	public function get_categories() {
		return [
			Module::IMAGE_CATEGORY,
			Module::MEDIA_CATEGORY,
		];
	}

    /**
     * Register controls
     */
    protected function register_controls() {
        // Get all available image sizes
        $image_sizes = $this->get_image_sizes();

        $this->add_control(
            'image_size',
            [
                'label' => __( 'Image Size', 'runthings-dynamic-tag-image-size' ),
                'type' => Controls_Manager::SELECT,
                'default' => 'full',
                'options' => $image_sizes,
                'description' => __( 'Select the image size to use. Temporary (?) workaround as default featured image tag only supports full size. <a href="
https://github.com/orgs/elementor/discussions/22750" target="_blank">GitHub issue #22750</a>', 'runthings-dynamic-tag-image-size' ),
            ]
        );

        $this->add_control(
            'fallback',
            [
                'label' => __( 'Fallback Image', 'runthings-dynamic-tag-image-size' ),
                'type' => Controls_Manager::MEDIA,
                'description' => __( 'Image to use if no featured image is set', 'runthings-dynamic-tag-image-size' ),
            ]
        );
    }

    /**
     * Get available image sizes
     */
    private function get_image_sizes() {
        $sizes = [];

        // Get all registered image sizes including dimensions
        global $_wp_additional_image_sizes;

        // Add full size
        $sizes['full'] = __( 'Full Size (original)', 'runthings-dynamic-tag-image-size' );

        // Get WordPress default sizes
        $default_sizes = [
            'thumbnail' => __( 'Thumbnail', 'runthings-dynamic-tag-image-size' ),
            'medium' => __( 'Medium', 'runthings-dynamic-tag-image-size' ),
            'medium_large' => __( 'Medium Large', 'runthings-dynamic-tag-image-size' ),
            'large' => __( 'Large', 'runthings-dynamic-tag-image-size' ),
        ];

        foreach ( $default_sizes as $size => $label ) {
            $width = get_option( $size . '_size_w' );
            $height = get_option( $size . '_size_h' );

            if ( $width && $height ) {
                $sizes[ $size ] = sprintf( '%s (%dx%d)', $label, $width, $height );
            } elseif ( $width && ! $height ) {
                $sizes[ $size ] = sprintf( '%s (%dpx wide)', $label, $width );
            } elseif ( ! $width && $height ) {
                $sizes[ $size ] = sprintf( '%s (%dpx tall)', $label, $height );
            } else {
                $sizes[ $size ] = $label;
            }
        }

        // Add additional registered image sizes
        if ( $_wp_additional_image_sizes ) {
            foreach ( $_wp_additional_image_sizes as $size => $data ) {
                $display_name = ucwords( str_replace( [ '_', '-' ], ' ', $size ) );
                $sizes[ $size ] = sprintf( '%s (%dx%d)', $display_name, $data['width'], $data['height'] );
            }
        }

        return $sizes;
    }

    /**
     * Get the value (image data array)
     */
    public function get_value( array $options = [] ) {
        $post_id = get_the_ID();
        
        if ( ! $post_id || ! has_post_thumbnail( $post_id ) ) {
            return $this->get_fallback_image();
        }

        $image_size = $this->get_settings( 'image_size' );
        $attachment_id = get_post_thumbnail_id( $post_id );

        if ( ! $attachment_id ) {
            return $this->get_fallback_image();
        }

        // Get image data for the specified size
        $image_data = wp_get_attachment_image_src( $attachment_id, $image_size );

        if ( ! $image_data ) {
            return $this->get_fallback_image();
        }

        // Return in the format Elementor expects
        return [
            // skip id or elementor loads from that instead of the url
            // 'id' => $attachment_id,
            'url' => $image_data[0],
            'width' => isset( $image_data[1] ) ? $image_data[1] : '',
            'height' => isset( $image_data[2] ) ? $image_data[2] : '',
        ];
    }

    /**
     * Get fallback image data
     */
    private function get_fallback_image() {
        $fallback = $this->get_settings( 'fallback' );
        
        if ( empty( $fallback['id'] ) ) {
            return [];
        }

        $image_size = $this->get_settings( 'image_size' );
        $image_data = wp_get_attachment_image_src( $fallback['id'], $image_size );

        if ( ! $image_data ) {
            return $fallback;
        }

        return [
            // skip id or elementor loads from that instead of the url
            // 'id' => $fallback['id'],
            'url' => $image_data[0],
            'width' => isset( $image_data[1] ) ? $image_data[1] : '',
            'height' => isset( $image_data[2] ) ? $image_data[2] : '',
        ];
    }
}

```

And it needs registering like:

```php
class Manager {
    /**
     * Initialize dynamic tags
     */
    public function __construct() {
        add_action( 'elementor/dynamic_tags/register', [ $this, 'register' ] );
    }
    /**
     * Register dynamic tags
     */
    public function register_dynamic_tags( $dynamic_tags ) {
        // Include tag classes
        require_once __DIR__ . '/FeaturedImageSizeTag.php';

        // Register tags
        $dynamic_tags->register( new FeaturedImageSizeTag() );
    }
```

Do it in two separate files so that the Elementor code is never included if the elementor action doesn't fire, which will stop the plugin breaking the site in the case that you delete your elementor plugin for some reason.

**Comment by zagarskas** [comment]
yay. good sir. yay.
we needs this. 
the yays have it
